### PR TITLE
NFT Mass Drop

### PR DESCRIPTION
**Pull request by JasonTulp**
Draft PR for NFT Mass Drop System
More tests and finalisations to come

### CODE REVIEW

File: crml/nft/src/lib.rs [hunk 1]
```diff
@@ -152,6 +164,34 @@ decl_error! {
...
 		/// Tokens with different individual royalties cannot be sold together
 		RoyaltiesProtection,
+		/// The mass_drop activation_time or limit is invalid
+		MassDropInvalid,
+		/// There is no mass_drop associated with the series
+		NoMassDropExists,
+		/// There are no accounts in the whitelist
+		WhitelistEmpty,
+		/// The account_id isn't in the whitelist
+		NoAccountId,
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]: could re-use `NoPermission` for this
```suggestion
		NoAccountId,
```

File: crml/nft/src/lib.rs [hunk 2]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+			let serial_number = Self::next_serial_number(collection_id, series_id);
+			ensure!(
+				serial_number.checked_add(quantity).is_some(),
+				Error::<T>::NoAvailableIds
+			);
+			let mass_drop: MassDrop<T::AccountId> = Self::series_mass_drops(collection_id, series_id)
+				.ok_or(Error::<T>::NoMassDropExists)?;
+			let current_block = <frame_system::Module<T>>::block_number();
+
+			if current_block >= T::BlockNumber::from(mass_drop.activation_time) || is_owner{
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]:
```suggestion
			if current_block >= T::BlockNumber::from(mass_drop.activation_time) || is_owner {
```

File: crml/nft/src/lib.rs [hunk 3]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+				// MassDrop has started, proceed with mint
+				if mass_drop.whitelist != vec![] {
+					ensure!(mass_drop.whitelist.contains(&origin) || is_owner, Error::<T>::NotInWhitelist);
+				}
+				if let Some(transaction_limit) = mass_drop.transaction_limit {
+					ensure!(quantity <= transaction_limit, Error::<T>::PurchaseQuantityTooHigh);
+				}
+				ensure!(serial_number + quantity <= mass_drop.max_supply, Error::<T>::NotEnoughTokens);
+				Self::process_drop_payment(&origin, owner, mass_drop.price, quantity.clone(), mass_drop.asset_id)?;
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]:
```suggestion
				Self::process_drop_payment(&origin, owner, mass_drop.price, quantity, mass_drop.asset_id)?;
```

File: crml/nft/src/lib.rs [hunk 4]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+			}else {
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]:
```suggestion
			} else {
```

File: crml/nft/src/lib.rs [hunk 5]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+				// Check if the presale has started
+				let pre_sale: PreSale<T::AccountId> = mass_drop.pre_sale
+					.ok_or(Error::<T>::MassDropNotStarted)?;
+				ensure!(current_block >= T::BlockNumber::from(pre_sale.activation_time), Error::<T>::PresaleNotStarted);
+				if pre_sale.whitelist != vec![] {
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]:
```suggestion
				if !pre_sale.whitelist.is_empty() {
```

File: crml/nft/src/lib.rs [hunk 6]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+		/// If the whitelist is not empty, only whitelisted accounts will be able to join
+		#[weight = 100000]
+		#[transactional]
+		fn enter_mass_drop(
+			origin,
+			collection_id: CollectionId,
+			series_id: SeriesId,
+			quantity: TokenCount,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
```
Comment (jordy25519) [inline crml/nft/src/lib.rs:443]: recommend restructure this function a little
1) check if owner / massdrop / presale + extract common vars (activation_time, whitelist, price, max_supply)
2) do checks
3) do payment + mint

File: crml/nft/src/lib.rs [hunk 7]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+				ensure!(serial_number + quantity <= pre_sale.max_supply, Error::<T>::NotEnoughTokens);
+				Self::process_drop_payment(&origin, owner, pre_sale.price, quantity.clone(), mass_drop.asset_id)?;
+			}
+			Self::do_mint_additional(&origin, collection_id, series_id, serial_number, quantity)?;
+			Self::deposit_event(RawEvent::EnterMassDrop(collection_id, series_id, quantity, origin));
+			Ok(())
+		}
+
+		/// Change the activation time of an existing mass_drop
+		#[weight = 0]
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]: some weight like 100,000 should do

File: crml/nft/src/lib.rs [hunk 8]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+				.ok_or(Error::<T>::NoMassDropExists)?;
+			let current_block = <frame_system::Module<T>>::block_number();
+
+			ensure!(
+				current_block < T::BlockNumber::from(mass_drop.activation_time) && current_block < T::BlockNumber::from(new_activation_time),
+				Error::<T>::ActivationTimeInvalid
+			);
+			mass_drop.activation_time = new_activation_time;
+			ensure!(mass_drop.validate(), Error::<T>::MassDropInvalid);
+			SeriesMassDrops::<T>::mutate(collection_id, series_id, |i| *i = Some(mass_drop));
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]:
```suggestion
			SeriesMassDrops::<T>::insert(collection_id, series_id, mass_drop);
```
don't want to read again here, just insert (already read the value on line 416)

File: crml/nft/src/lib.rs [hunk 9]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+			}
+			Self::do_mint_additional(&origin, collection_id, series_id, serial_number, quantity)?;
+			Self::deposit_event(RawEvent::EnterMassDrop(collection_id, series_id, quantity, origin));
+			Ok(())
+		}
+
+		/// Change the activation time of an existing mass_drop
+		#[weight = 0]
+		#[transactional]
+		fn update_mass_drop_activation_time(
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]:
```suggestion
		fn set_mass_drop_activation_time(
```
for consistency with set_whitelist

File: crml/nft/src/lib.rs [hunk 10]
```diff
@@ -152,6 +164,34 @@ decl_error! {
...
+		/// There are not enough tokens left to purchase this amount
+		NotEnoughTokens,
+		/// The activation time has passed or is invalid
+		ActivationTimeInvalid,
+		/// This account is not registered in the whitelist
+		NotInWhitelist,
+		/// No owner exists for this collection id
+		NoCollectionOwner,
+		/// Can't mint additional when a mass drop is active
+		MassDropActive,
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]: consider changing this so it's not confused with activation time
```suggestion
		MassDropConfigured,
```

File: crml/nft/src/lib.rs [hunk 11]
```diff
@@ -402,11 +643,9 @@ decl_module! {
...
 			let origin = ensure_signed(origin)?;
 
-			ensure!(quantity > Zero::zero(), Error::<T>::NoToken);
-			ensure!(!Self::is_single_issue(collection_id, series_id), Error::<T>::AddToUniqueIssue);
-			let serial_number = Self::next_serial_number(collection_id, series_id);
-			ensure!(serial_number > Zero::zero(), Error::<T>::NoToken);
-			ensure!(serial_number.checked_add(quantity).is_some(), Error::<T>::NoAvailableIds);
+			if Self::series_mass_drops(collection_id, series_id).is_some() {
+				return Err(Error::<T>::MassDropActive.into());
+			}
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]:
```suggestion
			ensure!(Self::series_mass_drops(collection_id, series_id).is_none(), Error::<T>::MassDropActive)?;
```

File: crml/nft/src/lib.rs [hunk 12]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+
+			ensure!(
+				current_block < T::BlockNumber::from(mass_drop.activation_time) && current_block < T::BlockNumber::from(new_activation_time),
+				Error::<T>::ActivationTimeInvalid
+			);
+
+			pre_sale.activation_time = new_activation_time;
+			mass_drop.pre_sale = Some(pre_sale);
+			ensure!(mass_drop.validate(), Error::<T>::MassDropInvalid);
+			SeriesMassDrops::<T>::mutate(collection_id, series_id, |i| *i = Some(mass_drop));
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]: insert instead of mutate

File: crml/nft/src/lib.rs [hunk 13]
```diff
@@ -351,6 +579,19 @@ decl_module! {
...
+			// Create the mass_drop data
+			if let Some(mass_drop) = mass_drop {
+				ensure!(mass_drop.validate(), Error::<T>::MassDropInvalid);
+				let current_block = <frame_system::Module<T>>::block_number();
+				ensure!(
+					current_block < T::BlockNumber::from(mass_drop.activation_time),
+					Error::<T>::ActivationTimeInvalid
+				);
+				SeriesMassDrops::<T>::insert(collection_id, series_id, mass_drop);
+			}else {
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]:
```suggestion
			} else {
```

File: crml/nft/src/lib.rs [hunk 14]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+			ensure!(Self::collection_owner(collection_id) == Some(origin), Error::<T>::NoPermission);
+			let mut mass_drop: MassDrop<T::AccountId> = Self::series_mass_drops(collection_id, series_id)
+				.ok_or(Error::<T>::NoMassDropExists)?;
+			let mut pre_sale: PreSale<T::AccountId> = mass_drop.pre_sale
+				.ok_or(Error::<T>::NoPresaleExists)?;
+			let whitelist = account_ids;
+
+			pre_sale.whitelist = whitelist.clone();
+			mass_drop.pre_sale = Some(pre_sale);
+			SeriesMassDrops::<T>::mutate(collection_id, series_id, |i| *i = Some(mass_drop));
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]: insert

File: crml/nft/src/lib.rs [hunk 15]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+			mass_drop.whitelist = account_ids.clone();
+			SeriesMassDrops::<T>::mutate(collection_id, series_id, |i| *i = Some(mass_drop));
+			Self::deposit_event(RawEvent::UpdateMassDropWhitelist(collection_id, series_id, account_ids));
+			Ok(())
+		}
+
+		/// Add presale conditions to existing mass_drop or change presale values
+		#[weight = 0]
+		#[transactional]
+		fn update_mass_drop_presale(
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]:
```suggestion
		fn set_mass_drop_presale(
```

File: crml/nft/src/lib.rs [hunk 16]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+				.ok_or(Error::<T>::NoMassDropExists)?;
+
+			ensure!(pre_sale.validate(&mass_drop), Error::<T>::PresaleInvalid);
+			let current_block = <frame_system::Module<T>>::block_number();
+			ensure!(
+				current_block < T::BlockNumber::from(pre_sale.activation_time),
+				Error::<T>::ActivationTimeInvalid
+			);
+			mass_drop.pre_sale = Some(pre_sale.clone());
+			SeriesMassDrops::<T>::mutate(collection_id, series_id, |i| *i = Some(mass_drop));
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]: insert

File: crml/nft/src/lib.rs [hunk 17]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+			mass_drop.pre_sale = Some(pre_sale);
+			SeriesMassDrops::<T>::mutate(collection_id, series_id, |i| *i = Some(mass_drop));
+			Self::deposit_event(RawEvent::UpdatePresaleWhitelist(collection_id, series_id, whitelist));
+			Ok(())
+		}
+
+		/// Change the activation time of an existing pre_sale
+		#[weight = 0] // Change this to appropriate weighting
+		#[transactional]
+		fn update_pre_sale_activation_time(
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]:
```suggestion
		fn set_pre_sale_activation_time(
```

File: crml/nft/src/types.rs [hunk 18]
```diff
@@ -87,6 +88,70 @@ pub struct TokenInfo<AccountId> {
 	pub royalties: Vec<(AccountId, Permill)>,
 }
 
+/// Contains information for a particular token. Returns the attributes and owner
```
Comment (jordy25519) [inline crml/nft/src/types.rs]:
```suggestion
/// Configuration for a mass drop 
```

File: crml/nft/src/lib.rs [hunk 19]
```diff
@@ -801,6 +1040,46 @@ impl<T: Config> Module<T> {
...
+			!Self::is_single_issue(collection_id, series_id),
+			Error::<T>::AddToUniqueIssue
+		);
+
+		// Mint the set tokens
+		for serial_number in serial_number..serial_number + quantity {
+			<TokenOwner<T>>::insert((collection_id, series_id), serial_number as SerialNumber, &owner);
+		}
+
+		SeriesIssuance::mutate(collection_id, series_id, |q| *q = q.saturating_add(quantity));
```
Comment (jordy25519) [inline crml/nft/src/lib.rs:1135]: this is a good example of where mutate is used well

File: crml/nft/src/lib.rs [hunk 20]
```diff
@@ -302,10 +344,195 @@ decl_module! {
...
+			collection_id: CollectionId,
+			series_id: SeriesId,
+			account_ids: Vec<T::AccountId>,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			ensure!(Self::collection_owner(collection_id) == Some(origin), Error::<T>::NoPermission);
+			let mut mass_drop: MassDrop<T::AccountId> = Self::series_mass_drops(collection_id, series_id)
+				.ok_or(Error::<T>::NoMassDropExists)?;
+			mass_drop.whitelist = account_ids.clone();
+			SeriesMassDrops::<T>::mutate(collection_id, series_id, |i| *i = Some(mass_drop));
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]: insert

File: crml/nft/src/types.rs [hunk 21]
```diff
@@ -87,6 +88,70 @@ pub struct TokenInfo<AccountId> {
...
+		if let Some(pre_sale) = &self.pre_sale {
+			if !pre_sale.validate(self) {
+				return false;
+			}
+		}
+		self.max_supply > 0
+	}
+}
+
+/// Contains information for a particular token. Returns the attributes and owner
```
Comment (jordy25519) [inline crml/nft/src/types.rs]: update comment

**Review by jordy25519 — COMMENTED** [review]
one thing to figure out is how to expire the whitelists / remove from state at a later time.
could bump into some scaling issues with large whitelists (searching and reading from storage)

File: crml/nft/src/lib.rs [hunk 22]
```diff
@@ -152,6 +164,34 @@ decl_error! {
 		AddToUniqueIssue,
 		/// Tokens with different individual royalties cannot be sold together
 		RoyaltiesProtection,
+		/// The mass_drop activation_time or limit is invalid
+		MassDropInvalid,
+		/// There is no mass_drop associated with the series
+		NoMassDropExists,
+		/// There are no accounts in the whitelist
+		WhitelistEmpty,
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]: I think this unused

File: crml/nft/src/lib.rs (hunk 19 above)
Comment (JasonTulp) [reply crml/nft/src/lib.rs:1135]: Oooooh I get it, so just replacing the value entirely we use insert, but if we need to read and write mutate. Like you've been saying this whole time 😅

File: crml/nft/src/lib.rs [hunk 23]
```diff
@@ -302,10 +340,174 @@ decl_module! {
...
+			let quantity = if serial_number + quantity >= max_supply {
+				PresaleWhitelist::<T>::remove_prefix((collection_id, series_id));
+				max_supply - serial_number
+			}else {
+				quantity
+			};
+
+			Self::process_drop_payment(&origin, owner, price, quantity, mass_drop.asset_id)?;
+			Self::do_mint_additional(&origin, collection_id, series_id, serial_number, quantity)?;
+			Self::deposit_event(RawEvent::EnterMassDrop(collection_id, series_id, quantity, origin));
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]: please add the first serial number in this event
```suggestion
			Self::deposit_event(RawEvent::EnterMassDrop(collection_id, series_id, quantity, origin));
```

File: crml/nft/src/lib.rs [hunk 24]
```diff
@@ -302,10 +340,174 @@ decl_module! {
...
+				return Err(Error::<T>::MassDropNotStarted.into());
+			};
+
+			if let Some(transaction_limit) = transaction_limit {
+				ensure!(quantity <= transaction_limit, Error::<T>::PurchaseQuantityTooHigh);
+			}
+
+			ensure!(serial_number <= max_supply, Error::<T>::NotEnoughTokens);
+			let quantity = if serial_number + quantity >= max_supply {
+				PresaleWhitelist::<T>::remove_prefix((collection_id, series_id));
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]: could drop a presale ended event here

File: crml/nft/src/lib.rs [hunk 25]
```diff
@@ -302,10 +340,174 @@ decl_module! {
...
+				max_supply - serial_number
+			}else {
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]: it's annoying but `cargo fmt` doesn't process stuff in macro blocks atm
```suggestion
			} else {
```

File: crml/nft/src/lib.rs [hunk 26]
```diff
@@ -801,6 +1012,46 @@ impl<T: Config> Module<T> {
...
+
+		SeriesIssuance::mutate(collection_id, series_id, |q| *q = q.saturating_add(quantity));
+		NextSerialNumber::mutate(collection_id, series_id, |q| *q = q.saturating_add(quantity));
+
+		Ok(())
+	}
+
+	fn process_drop_payment(
+		buyer: &T::AccountId,
+		owner: T::AccountId,
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]:
```suggestion
		owner: &T::AccountId,
```

File: crml/nft/src/tests.rs [hunk 27]
```diff
@@ -1781,3 +1850,897 @@ fn get_collection_info() {
...
+		};
+		let mass_drop = MassDrop {
+			price: 20,
+			asset_id: 16000,
+			max_supply: 1000,
+			transaction_limit: Some(1001),
+			activation_time: 10,
+			presale: None,
+		};
+		//Activation time <= current block should fail
```
Comment (jordy25519) [inline crml/nft/src/tests.rs]: update comment

File: crml/nft/src/tests.rs [hunk 28]
```diff
@@ -1781,3 +1850,897 @@ fn get_collection_info() {
...
+		};
+		let mass_drop = MassDrop {
+			price: 20,
+			asset_id: 16000,
+			max_supply: 0,
+			transaction_limit: None,
+			activation_time: 10,
+			presale: None,
+		};
+		//Activation time <= current block should fail
```
Comment (jordy25519) [inline crml/nft/src/tests.rs:2004]: update comment

File: crml/nft/src/tests.rs [hunk 29]
```diff
@@ -1781,3 +1850,897 @@ fn get_collection_info() {
...
+
+		assert_noop!(
+			Nft::enter_mass_drop(Some(2_u64).into(), fixture.collection_id, fixture.series_id, quantity,),
+			Error::<Test>::MassDropNotStarted
+		);
+	});
+}
+
+#[test]
+fn enter_presale_on_whitelist() {
```
Comment (jordy25519) [inline crml/nft/src/tests.rs:2696]: where is the whitelist configured in this test?

**Review by jordy25519 — APPROVED** [review]
lgtm, couple nits to tidy up

File: crml/nft/src/tests.rs (hunk 29 above)
Comment (jordy25519) [reply crml/nft/src/tests.rs:2696]: want to make sure we're testing the presale with whitelist success case

File: crml/nft/src/lib.rs [hunk 30]
```diff
@@ -348,6 +356,46 @@ decl_module! {
...
+						Error::<T>::MetadataAlreadySet
+					);
+				}
+			}
+			// Loop through each metadata uri
+			for (i, uri) in metadata_uris.iter().enumerate() {
+				let serial_number: SerialNumber = (i as u32).into();
+				SerialNumberMetadataURI::insert((collection_id, series_id), serial_number, uri);
+			}
+			Self::deposit_event(RawEvent::MetadataUpdated(collection_id, series_id));
```
Comment (jordy25519) [inline crml/nft/src/lib.rs]: could you add the last serial number in the event (helps scanners)

**Review by jordy25519 — APPROVED** [review]

**Comment by jordy25519** [comment]
mass merge

**Pull request merged**